proposals/010-authorizer.md

```diff
@@ -0,0 +1,212 @@
+<!-- This template is provided as an example with sections you may wish to comment on with respect to your proposal. Add or remove sections as required to best articulate the proposal. -->
+
+# Authorizer Filter
+
+The Authorizer filter gives the ability to add authorisation checks into a Kafka system which will be enforced by the proxy. 
+
+## Current situation
+
+It is possible for a filter to implement its own business rules, enforcing authorization in some custom manner.  However,
+that approach does not have good separation of concerns.  Authorization checks are an orthogonal concern, and security
+best practice is to separate their enforcement from business logic.
+
+## Motivation
+
+We are identifying use-cases where making authorization decisions at the proxy is desirable.   Examples include where one wishes to restrict a virtual cluster to a sub-set of the resources (say topics) of the cluster.
+
+## Proposal
+
+The Authorizer filter gives the ability to layer authorization checks into a Kafka system which with those authorization checks being enforced by the filter.  These authorization checks are in addition to anythat may be imposed by the Kafka Cluster itself. This means that for an action to be allowed both the proxy’s authorizer and the Kafka broker’s authorizer will need to reach an ALLOW decision.
+
+The Authorizer filter allows for authorization checks to be made in the following form:
+
+`Principal P is [Allowed/Denied] Operation O On Resource R`.
+
+where:
+
+* Principal is the authenticated user.
+* Operation is an action such as, but not limited to, Read, Write, Create, Delete.
+* Resource identifies one or more resources, such as, but not limited to Topic, Group, Cluster, TransactionalId.
+
+Unlike Apache Kafka authorizer system, the `from host` predicate is omitted.  This is done to adhere to the modern security principle that there are no privileged network locations.
+
+### Request authorization
+
+The Authorizer filter will intercept all request messages that perform an action on a resource, and all response messages that list resources.  
+
+On receipt of a request message from the downstream, the filter will make an asynchronous call to the authorizer for the resource(s) involved in the request.  If the authorization result for all resources is `ALLOWED`, the filter will forward the request to the broker.
+If the authorization result is `DENIED` for any resource in the request, the filter will produce a short circuit error response denying the request using the appropriate authorization failed error code.  The authorizer filter must not forward requests that fail authorization. 
+
+### Response resource filtering
+
+On receipt of a response message from the upstream, the Authorizing filter will filter the resources so that the downstream receives only resources that they are authorized to `DESCRIBE`.
```

Comment:
Something Tom pointed out in discussion is that Metadata, DescribeGroups, DescribeCluster, ConsumerGroupDescribe, ShareGroupDescribe responses can contain authorized operations bitmasks. Will this Filter manipulate the metadata for these RPCS? Should we be generating our own mask for these resources and or-ing it with the upstream?

Reply:
Agreed, we have to implement this correctly.